### PR TITLE
feat: process genesis

### PR DESCRIPTION
**Pull request by bryanchriswhite**
## Apology

My apologies for the very large PR. Typically I make more of an effort to split things up into smaller pieces (PRs) so that review is not such a daunting task. However, I'm making a rare exception In the interest of overall expediency as this work is already overdue.

## Suggestion

I've taken some time to reorganize the commits on this branch to represent logical steps towards the end goal. It might be easier to review these commits individually and leave comments on the corresponding line(s) in the "files changed" tab.

## Changes

- factor sql and gql clients out of base test case
- reorganize tests
- add genesis
- add state classes
- add reactivex dependency
- add chain-id to account schema
- add table manager
- add chain id observer
- add accounts observer and manager
- add balances observer and manager
- add genesis processing script
- tidy test imports for usage with unittest discover
- update test workflow to run e2e and unit tests

### CODE REVIEW

File: src/genesis/genesis.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+
+@dataclass
+class PubKey:
+    type: str
+    value: str
+
+
+@dataclass
+class ValidatorData:
```
Comment (bryanchriswhite) [inline src/genesis/genesis.py:21]: Add defaults

File: src/genesis/genesis.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    address: str
+    name: str
+    power: str
+    pub_key: PubKey
+
+
+class Validator(ValidatorData):
+    def __init__(self, **kwargs):
+        kwargs["pub_key"] = PubKey(**kwargs["pub_key"])
```
Comment (bryanchriswhite) [inline src/genesis/genesis.py:30]: Check if key in kwargs

File: src/genesis/state/bank.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+@dataclass(frozen=True)
+class BalanceData(ListConstructorMixin):
+    address: str
+    coins: List[Coin]
+
+
+class Balance(BalanceData, ListConstructorMixin, OwnAttrsMixin):
+    def __init__(self, **kwargs):
+        kwargs["coins"] = Coin.from_dict_list(kwargs["coins"])
```
Comment (bryanchriswhite) [inline src/genesis/state/bank.py:15]: Check if key in kwargs

File: src/genesis/state/bank.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
+from dataclasses import dataclass, field
+from typing import List, Dict
+
+from src.genesis.state.common import Coin, ListConstructorMixin, list_field_with_default, OwnAttrsMixin
+
+
+@dataclass(frozen=True)
+class BalanceData(ListConstructorMixin):
```
Comment (bryanchriswhite) [inline src/genesis/state/bank.py:8]: Add defaults

File: src/genesis/state/distribution.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
+from dataclasses import dataclass
+from typing import List, Dict
+
+from src.genesis.state.common import Coin, ListConstructorMixin
+
+
+@dataclass(frozen=True)
+class StartingInfo:
```
Comment (bryanchriswhite) [inline src/genesis/state/distribution.py:8]: Add defaults

File: src/genesis/state/distribution.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    params: DistParams
+    previous_proposer: str
+    validator_accumulated_commissions: List[ValidatorAccumulatedCommission]
+    validator_current_rewards: List[ValidatorCurrentReward]
+    validator_historical_rewards: List[ValidatorHistoricalReward]
+    validator_slash_events: List[ValidatorSlashEvent]
+
+    def __init__(self, **kwargs: Dict[str, any]):
+        concrete_args = {
+            "delegator_starting_infos": DelegatorStartingInfo.from_dict_list(kwargs["delegator_starting_infos"]),
```
Comment (bryanchriswhite) [inline src/genesis/state/distribution.py:105]: Use `Dict#get`

File: src/genesis/state/gov.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
+from dataclasses import dataclass
+from typing import List, Dict
+
+from src.genesis.state.common import Coin
+
+
+@dataclass(frozen=True)
+class DepositParams:
```
Comment (bryanchriswhite) [inline src/genesis/state/gov.py:8]: Add defaults

File: tests/helpers/utils.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+def check_attrs(test_case: unittest.TestCase, expected: any, actual: any):
+    for attr in actual.attrs():
+        # TODO: unworkaround
+        if attr == "data":
+            expected_attr = expected
+        else:
+            expected_attr = expected[attr]
+        actual_attr = getattr(actual, attr)
+
+        def check(expected_: any, actual_: any):
```
Comment (bryanchriswhite) [inline tests/helpers/utils.py:31]: Promote these to package-level

File: schema.graphql [hunk 9]
```diff
@@ -136,8 +136,17 @@ type Cw20Transfer @entity {
 type Account @entity {
   # id is the address
   id: ID!
+  chainId: String!
```
Comment (bryanchriswhite) [inline schema.graphql:139]:
```suggestion
  chainId: String! @index
```

**Review by Jonathansumner — APPROVED** [review]
Finally got through it! Looks great Bryan, nice work

**Review by Jonathansumner — APPROVED** [review]
LGTM!

**Pull request merged**